### PR TITLE
Easygpio: module for reading bits from sensors

### PR DESCRIPTION
**Pull request by romanchyla**
- easygpio: generic bit-banging
- added documentation
- added short tutorial for writing a c module for nodemcu

I've taken the opportunity to provide a short tutorial. I remember how hard it was when I tried to write my first module few months back (at that time, one had to edit 4 files). And for beginners it is really a very steep hill to climb.

Please consider it as suggestion, the easygpio module doesn't have to be enabled by default. But I'm hoping it could be useful.

### CODE REVIEW

**Comment by marcelstoer** [comment]
Here's my 2 cents:
- Don't include unrelated changes in a single PR. `cmodule.md` is nice but should be in its own PR (along with the new FAQ entry).
- New module should be disabled by default in `user_modules.h` and not be inserted at the very end (modules are alphabetically ordered).
- Link to new module documentation must be added to `mkdocs.yml`, see https://github.com/nodemcu/nodemcu-firmware/blob/dev/CONTRIBUTING.md#writing-documentation.

**Comment by TerryE** [comment]
My immediate reaction is that this is really the equivalent of gpio.serout, but for input.  It makes sense to combine these two sets into one module.  See #428 for some of the difficulties here.  As these are a niche use, I suspect that these would be better in a separate module -- but not "easy"gpio.  However, we are in the middle of tidying up / reworking the gpio implementation, so do you mind if we keep this on hold for a few weeks?

**Comment by romanchyla** [comment]
Thank you for suggestions, I'll split the PR into two parts - I belive the cmodule.md tutorial is uncontroversial (and for beginners, it would be great to have it); I'll make a separate PR for it.

@TerryE I also thought it better to keep the functionality in a separate module (maybe also serout could be extracted?). Can I somehow help with the gpio module itself? [ie: I'd like to avoid having to track several repositories for my project which depend on the bit-banging]

**Comment by romanchyla** [comment]
I've changed the PR as indicated above. If you would like me to do something else or help with the gpio, please let me know. Thank you.

File: app/modules/gpioser.c [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+long read(int pin_clk, int pin_data, int num_bits, int after_ticks,
+        int delay_us, int initial_clk_write, int ready_state) {
+    unsigned long Count;
+    unsigned char i;
+    Count=0;
+
+    platform_gpio_write(pin_clk, initial_clk_write);
+
+    while (platform_gpio_read(pin_data) != ready_state);
```
Comment (pjsg) [inline app/modules/gpioser.c]: This can hang forever if the device doesn't respond (well, the watchdog will kill it eventually).

**Comment by marcelstoer** [comment]
@romanchyla pls address Philip's comment, rebase on current `dev` to fix the merge conflicts and then push to your branch again.

**Comment by romanchyla** [comment]
ok, will do - give me few days...

**Comment by romanchyla** [comment]
ok, updated - it is trying 1000 times before giving up (not sure that is the right solution, but can't think of anything else&simple: is that bad/low/too arbitrary?)

**Comment by pjsg** [comment]
I looked at the code, and I don't see any assignment to PIN_CLK (the global variable). 

It also isn't clear that the info() method provides much value....

File: docs/en/modules/gpioser.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+This module can be used for reading bits from wide range of sensors (bit banging).
+
+## gpioser.read()
+Reads data from the GPIO sensors using specified intervals. 
+
+#### Syntax
+`gpioser.read(pin_clk, pin_out, num_bits, after_ticks, delay_us, initial_val, ready_state)`
+
+#### Parameters
+`pin_clk` pin number of the sensor (the CLOCK channel), type is number
```
Comment (marcelstoer) [inline docs/en/modules/gpioser.md]: All the parameters are missing the list markers ('- ').

**Comment by marcelstoer** [comment]
Thanks Roman for the update.

> this is really the equivalent of [`gpio.serout`](http://nodemcu.readthedocs.org/en/dev/en/modules/gpio/#gpioserout), but for input. It makes sense to combine these two sets into one module

So, `gpio.serout` could be transferred to this new module even if #428 isn't addressed yet. Or the other way around...The more I think about this the more I tend to the latter. It's not that `gpio.c` would grow beyond a manageable size if it also included a `serin` function.

> It also isn't clear that the info() method provides much value

I have to agree.

**Comment by romanchyla** [comment]
I vaguely remember looking at the gpio.serout (seems more flexible), the
code was hard to read though. It received a table with wait times. It will
be slower but how much (and if that matters) I don't know. Will try to do
some work on it next week.

For my edification: any idea how long the gpio_read() takes? And if it's
safe to call it 1000 times?

**Comment by marcelstoer** [comment]
What a pity this has been stuck in the PR queue for so long. Terry/Philip would it be unfair argue it should be closed as no one seems to have come forward in the mean time stating they really missed what this PR would provide?

**Comment by TerryE** [comment]
> Terry/Philip would it be unfair argue it should be closed as no one seems to have come forward in the mean time stating they really missed what this PR would provide?

It is also pretty out of date and would need work to make it compatible with current builds.   Time to let it die a natural death.

**Pull request closed without merging**